src/main/core/support/configuration.rs

```diff
@@ -953,6 +953,89 @@ impl<'de> serde::Deserialize<'de> for ProcessArgs {
     }
 }
 
+#[derive(Debug, Copy, Clone)]
+pub struct Signal(nix::sys::signal::Signal);
+
+impl serde::Serialize for Signal {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_str(self.0.as_str())
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for Signal {
+    fn deserialize<D: serde::Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        struct SignalVisitor;
+
+        impl<'de> serde::de::Visitor<'de> for SignalVisitor {
+            type Value = Signal;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                formatter.write_str("a signal string or integer")
+            }
+
+            fn visit_string<E>(self, v: String) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                self.visit_str(&v)
+            }
+
+            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                nix::sys::signal::Signal::from_str(v)
+                    .map(Signal)
+                    .map_err(|_e| E::custom(format!("Invalid signal string: {v}")))
+            }
+
+            fn visit_i64<E>(self, v: i64) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                let v = i32::try_from(v)
+                    .map_err(|_e| E::custom(format!("Invalid signal number: {v}")))?;
+                nix::sys::signal::Signal::try_from(v)
+                    .map(Signal)
+                    .map_err(|_e| E::custom(format!("Invalid signal number: {v}")))
+            }
+
+            fn visit_u64<E>(self, v: u64) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                let v = i64::try_from(v)
+                    .map_err(|_e| E::custom(format!("Invalid signal number: {v}")))?;
+                self.visit_i64(v)
+            }
+        }
+
+        deserializer.deserialize_any(SignalVisitor)
+    }
+}
+
+impl JsonSchema for Signal {
+    fn schema_name() -> String {
+        String::from("Signal")
+    }
+
+    fn json_schema(_gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
+        // TODO: is there a way to generate a schema for "string or integer"?
+        schemars::schema::Schema::Bool(true)
+    }
```

Comment:
I'm not sure the proper way to construct the schema object, but in Shadow we only read the `description` and the `default` fields of the metadata in order to output cli help messages. Since the description and default should come come from the parent structure's `shutdown_signal` field (`Signal` doesn't implement `Default`, so the default has to be defined by `ProcessOptions`), I don't think anything used here will affect the cli help messages.

But we only uses this JSON schema information for options that can be set on the cli, but `shutdown_signal` is not configurable through the cli. So we don't really need schema information for this field anyways (or really the entire `ProcessOptions` or `HostOptions` I think). So alternatively you could add `#[schemars(skip)]` to `shutdown_signal` or remove the `JsonSchema` derive from all of `Signal`, `ProcessOptions`, and `HostOptions`.

Reply:
Thanks - updated this comment.